Persist additional fields and payment method on checkout

alexflorisca opened the pull request:
### Changes proposed in this Pull Request:

This PR persists the additional fields tied to an order, as well as the order notes and payment method on checkout. This means all checkout fields are now persisted, and enables developers to do things like changing the cart totals or adding fees based on any checkout field.

Notable changes

- new `PUT` /checkout endpoint that persists the fields and updates totals. This returns a cart, which is updated client side
- push-changes, similar to the cart store, for the checkout store that sync with server.
- PUT requests get cancelled when clicking Place Order
- Note that saved payment methods are not persisted as part of this PR due to complications with tokens. We can address this in a future PR if necessary.

Closes #42079.

<!-- Begin testing instructions -->

### How to test the changes in this Pull Request:

<!-- Include detailed instructions on how these changes can be tested. Review and follow the guide for how to write high-quality testing instructions. -->

Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions:

**Pre-testing**
1. Install and activate the [Additional Checkout Fields Tester](https://github.com/woocommerce/additional-checkout-fields-tester) plugin
2. Install and activate the [Stripe Payment Gateway](https://wordpress.org/plugins/woocommerce-gateway-stripe/) plugin
3. Enable the Direct bank transfer, cheque, cash on delivery and stripe payment methods. 
4. Make sure you have a saved payment method available for your user. You can do this by checking out with stripe once, while making sure the checkbox `Save payment information to my account for future purchases.` is ticked.

**Test the Cart page**
1. Add some items to your cart & visit the cart page
2. Make sure there is not PUT request on page load
3. Click on an express payment method and close the modal. There should be no PUT request.

> [!NOTE]
>  The following test scenarios all happen on the Checkout page, unless otherwise mentioned.

**Test the additional fields on checkout**
1. Go to checkout
2. Make sure there is no PUT request on page load
3. Click an express payment method and close the modal. Make sure there are no PUT requests.
5. Fill in each additional field under contact information and additional information
6. Make sure there is a PUT request to the `wc/store/checkout` endpoint for each one
7. Make sure there is no PUT request to `wc/store/checkout`  endpoint when the email is filled in or the create account checkbox is ticked.
8. Refresh the page. The information you entered should still be there.

**Test the order notes**
1. Fill in the order notes and again check for a PUT request.
2. Refresh the page and make sure the data has persisted.
3. Untick the Order notes checkbox and refresh the page again.
4. The order notes checkbox should not be ticked. When ticking the checkbox again, the text field should be empty.

**Test the payment methods**
1. With a saved payment method active, change the payment method to something else, wait 1.5s and refresh
2. The saved payment method should be selected.
3. Disable the gateway with the saved payment option.
4. Change the payment method, wait 1.5s and refresh again
5. The payment method you selected should be selected.

**Test there are no regressions with the address fields**
1. Edit the address fields and check they are persisted by refreshing the page.

**Test that totals update based on changes to persisted fields**
1. Insert [this code snippet](https://gist.github.com/alexflorisca/051e0fe9b0bae22f5c4efdae9c505ba9) in either Code Snippets plugin or somewhere like `woocommerce.php` or `functions.php`.
2. Make sure you have the following payment methods active: Direct debit, Cash on delivery and Cheque.
3. Now go to the Checkout and change the payment method.
6. You should see the total update after the fetch request.

**Misc**
1. In the alternative email field, enter an invalid email and make sure the following error is displayed
2. Fill in all fields with valid data and place the order
3. In wp-admin, head to WooCommerce > Orders and make sure all the additional fields and order notes are stored correctly on the order you just placed.

![Screenshot 2025-01-07 at 15 32 54](https://github.com/user-attachments/assets/070db702-7e64-4d27-8116-a8204902efb6)

**Test that in progress or queued PUT requests get cancelled**

> [!NOTE]
>  Note these can be tested better by changing some code (see developer testing) but the following will do as a smoke test

1. Make sure all required fields are filled in on the Checkout form.
2. You'll have to be quick here. With your network tab open, enter some order notes and hit the place order button as fast as you can after. There is a 1.5s delay between finishing typing and the fetch request happening. You should not see any PUT requests after you have clicked Place Order.
3. Add some more items to your cart and go back to the checkout .
4. Open the network tab and toggle your network speed to the slowest possible. In Chrome you can add a custom profile. 1kbps works best here.
7. Again with your network tab open, make a change to the order notes. This time wait for the fetch request to `/wc/store/v1/checkout?calc=true`.
8. Once you see the request pending, click the Place Order button as soon as you can.
9. Check that the PUT request is cancelled.

**Developer testing only**

1. Remove this line to throw an error in https://github.com/woocommerce/woocommerce/blob/add/42079-additional-fields%3Dpersistent-updates/plugins/woocommerce/src/StoreApi/Routes/V1/Checkout.php#L311
2. Check that the error is displayed on the Checkout.

**Test that in progress or queued PUT requests get cancelled**

This is a more reliable way to check that the PUT requests get cancelled when placing the order

1. Increase the delay [here](https://github.com/woocommerce/woocommerce/blob/add/42079-additional-fields%3Dpersistent-updates/plugins/woocommerce-blocks/assets/js/data/checkout/push-changes.ts#L111) to something like 5
2. Open network tab, edit the order notes, and click place the order before the delay. No PUT request should appear.
3. Put a `sleep(10)` [here](https://github.com/woocommerce/woocommerce/blob/add/42079-additional-fields%3Dpersistent-updates/plugins/woocommerce/src/StoreApi/Routes/V1/Checkout.php#L273)
4. Open the tab, edit the order notes and while the fetch is ongoing, click the Place Order button. The PUT should be cancelled.


<!-- End testing instructions -->

### Changelog entry

<!-- You can optionally choose to enter a changelog entry by checking the box below and supplying data. -->
<!-- It will trigger the 'Add changelog to PR' CI job to create and push the entry into the branch. -->

<!-- Due to org permissions, the job might fail for PRs crated from a fork under GitHub organizations. Possible solutions: -->
<!-- * Create entry manually with `pnpm --filter='@woocommerce/plugin-woocommerce' changelog add` and push it into the branch (replace `@woocommerce/plugin-woocommerce` with package name from nearest `package.json` file) -->
<!-- * Create entry from supplied PR data and push it automatically `pnpm utils changefile pr-number-here -o github-org-name-here` -->

-   [x] Automatically create a changelog entry from the details below.

<!-- If no changelog entry is required for this PR, you can specify that below and provide a comment explaining why. This cannot be used if you selected the option to automatically create a changelog entry above. -->

-   [ ] This Pull Request does not require a changelog entry. (Comment required below)

<details>

<summary>Changelog Entry Details</summary>

#### Significance

<!-- Choose only one -->

-   [ ] Patch
-   [x] Minor
-   [ ] Major

#### Type

<!-- Choose only one -->

-   [ ] Fix - Fixes an existing bug
-   [x] Add - Adds functionality
-   [ ] Update - Update existing functionality
-   [ ] Dev - Development related task
-   [ ] Tweak - A minor adjustment to the codebase
-   [ ] Performance - Address performance issues
-   [ ] Enhancement - Improvement to existing functionality

#### Message <!-- Add a changelog message here -->

Additional checkout fields, order notes and chosen payment method are now persisted.

</details>

<details>

<summary>Changelog Entry Comment</summary>

#### Comment <!-- If your Pull Request doesn't require a changelog entry, a comment explaining why is required instead -->

</details>

Commit:
PUT changes to additional fields

Commit:
Remove console logs

Commit:
PUT endpoint

Commit:
WIP

Commit:
Save the order

Commit:
push changes for checkout store & persist order notes

Commit:
Persist payment method

Commit:
fix a couple of bugs with push-changes

Commit:
Return and sync cart with data store

Commit:
Handle error responses and receiveCart

Commit:
Update block.tsx

Commit:
Update block.tsx

Commit:
Remove duplicate processErrorResponse

alexflorisca commented on plugins/woocommerce/src/StoreApi/Routes/V1/Checkout.php:
```diff
@@ -198,7 +228,15 @@ public function validate_required_additional_fields( \WP_REST_Request $request )
+	 * @param \WP_REST_Request $request Request object.
+	 * @throws RouteException When a required address field is missing.
+	 */
+	private function validate_required_additional_fields_for_customer( \WP_REST_Request $request ) {
```
I split the `validate_required_additional_fields` function into two here, to be able to validate customer and order fields separately.

Commit:
remove unused namespace

Commit:
Add changefile(s) from automation for the following project(s): woocommerce-blocks, woocommerce

Commit:
Merge branch 'trunk' into add/42079-additional-fields=persistent-updates

senadir reviewed (commented):
Initial notes are that:

- We should rename the `cart` to `__exprimentalCart` as we're not sure if we will commit to it and ideally can roll it back.
- To not bloat existing requests for other consumers like WooPay, who are sensitive to speed, can we guard the calculation and fetching of cart to behind flag? sending a query param with the request is fine cc @mikejolley 

The page has a warning when loaded, you're not defining $cart, you should be using PHP 8+ to catch those warning

```
Warning: Undefined property: stdClass::$cart in /Users/nadir/work/monorepo/plugins/woocommerce/src/StoreApi/Schemas/V1/CheckoutSchema.php on line 213
```

Commit:
define cart

senadir reviewed (changes requested):
I started testing this PR with https://github.com/woocommerce/additional-checkout-fields-tester and there are already a couple of issues that we need to work around and fix:

- On cart, a request to `PUT /checkout` is being made to persist payment method, 2 issues here: 
    - Such code should not trigger on cart, I suspect this is because of the payment data store? Cart only has express payment methods, so such logic can maybe be excluded from there?
    - We're not passing down any or persisting a payment method on the sever when there's nothing in session, we need to match what's being for shipping methods in which the first one is persisted if nothing is available, @mikejolley can provide better direction here, let's see what shortcode is doing, if they're persisting something by default, and what are out options.
    - The same request is also happening in Checkout load, but I assume fixing point 1.2 should fix this as well.
- All additional fields are validated on `PUT /checkout`, this is the most glaring issue, we can't validate fields we didn't reach yet, as you're just reusing the same function in `POST` (`validate_required_additional_fields_for_order`), which at that point, it's valid to expect all fields to be present. For this, I think you should just create a separate function that validates passed fields only, do check `abstractAddressSchema.php` validate callback.
- For some reason (I did not dig too deep), `contact` fields are still throwing an error even when they have a value persisted, this is coming from the request that persists the payment method, do please check into that.

To summarize:
- [ ] There should be no extra requests on load, payment method should be selected server side.
- [ ] Partial validation should only happen for fields already persisted.

senadir commented:
On a separate note, I'm also getting a `Undefined property: stdClass::$payment_result in /Users/nadir/work/monorepo/plugins/woocommerce/src/StoreApi/Schemas/V1/CheckoutSchema.php on line 214
` warning, same as the `$cart` one.

mikejolley commented on plugins/woocommerce-blocks/assets/js/data/checkout/push-changes.ts:
```diff
@@ -0,0 +1,110 @@
+	localState.checkoutData = newCheckoutData;
+
+	dispatch( STORE_KEY )
+		.updateDraftOrder( {
```
Is checkout essentially blocked when this is taking place?

alexflorisca replied:
You can still interact with the UI, it happens in the background async

alexflorisca replied:
Should be the same as the cart fetch when we update the customer data

mikejolley commented on plugins/woocommerce-blocks/assets/js/data/checkout/thunks.ts:
```diff
@@ -135,3 +139,31 @@ export const __internalEmitAfterProcessingEvents: emitAfterProcessingEventsType
+		try {
+			const response = await apiFetchWithHeaders( {
+				path: '/wc/store/v1/checkout',
+				method: 'PUT',
```
👍🏻 I'm guessing validation on PUT needs to be more lax to prevent the route blocking partial updates.

alexflorisca replied:
what do you mean? Validation of additional fields server side?

senadir commented:
> I assume this means you need to wait for the server to update before you can place order e.g. if I switch to BACS payment method, I won't be able to place the order until the push is successful. Is this a concern?

While not implemented, I think it's fine to ignore any ongoing `PUT /checkout` requests if you make a `POST` one. The POST should be the final source of truth and it includes the whole state.

alexflorisca commented:
Thanks for your comments both.

> There are other ways to persist values without pushing to server, e.g. local storage, we could consider if it makes sense.
I think the benefit of server push here is that developers can also update the cart in response to changes to additional fields, payment method or order. But local storage would be a lot faster.

> While not implemented, I think it's fine to ignore any ongoing PUT /checkout requests if you make a POST one. The POST should be the final source of truth and it includes the whole state.

Agree with this. I think there's a danger here that A PUT request can be in progress, some cart totals get modified as a result (e.g. a plugin adding a fee), then the POST happens with updated totals that the user hasn't seen. Maybe disabling place order button while requests are in progress would work better?

Commit:
Rename cart to __experimentalCart

senadir commented:
I don't think a UI block is needed, unlike address update, those fields are unlikely to change totals. If we feel like we must do so, let's move this to be opt-in per phone.

alexflorisca commented:
> We should rename the cart to __exprimentalCart as we're not sure if we will commit to it and ideally can roll it back.

I don't think this is needed because it's only on the PUT endpoint. And it's needed to update totals there right? Or is that something we don't want to commit to yet?

Commit:
Only send the changed additional field on PUT

Commit:
Fix payment_result error

Commit:
Calculate totals behind a flag

Commit:
default payment on the server

alexflorisca commented:
Right, I think I addressed all of the concerns. 
- payment method is persisted and set to default if not set.
- default payment method is calculated and sent to the client. I tried to remove the whole set-default-payment-method client side code but I went down a rabbit hole of having to calculate and send the saved payment methods as well. It's doable, but it turned out more complicated than I thought so I'll write a follow up issue for that.
- changed `cart` to `__experimentalCart`
- Got rid of the warnings
- Removed validation of fields on PUT and only the field sent is persisted.
- Removed the push on cart.

@senadir Have another look and see what you think

alexflorisca commented:
Ah I still need to cancel the PUT request on POST. Will do that tomorrow.

Commit:
Clear any debounced PUT requests before POSTing

Commit:
Merge branch 'trunk' into add/42079-additional-fields=persistent-updates

alexflorisca commented:
I've added some functionality to clear PUT requests but it's not 100% there. Any debounced PUT requests that haven't yet been executed will be cancelled when the Place order button is clicked. However, because we're using batch, the standard way of using `AbortController` doesn't work for cancelling in flight requests. I didn't have too much time to big into the dataloader library that we use but I couldn't see any mention of how to do this in their docs. Any help here would be aprreciated.

Another challenge I ran into is with the payment method sync. I tried to avoid moving the saved payment logic server side as it's definitely needed on the client and feels weird to me to have it in both places. But seems without doing so, there will always be an extra fetch request (although only for signed in users with saved payment methods). Or the state will be out of sync between client and server, which is not ideal. At first I thought this might be ok because we're sending everything with the POST request, so that will have the updated payment method. However if we're allowing conditional logic based on payment method - adding fees for example - then if the server thinks the selected method is bacs, but the client has updated it to one of the user's saved methods, that could be problematic. 

I was hoping to wrap this up before I left for the holidays, but alas, here we are. If anyone has any ideas about the above let me know and I'll pick this back up when I can, maybe between a couple of turkey sandwiches to have a break from the inlaws.

Commit:
Persist saved payment methods

Commit:
couple of fixes

Commit:
Fix schema validation

Commit:
Tidy up

Commit:
Prevent PUT checkout endpoint from using batch so we can cancel requests

Commit:
Fix linting

Commit:
Add tests

alexflorisca commented:
This is ready for review again @senadir. I've made the checkout PUT endpoint not use batch so that we can cancel the requests, and I've added some logic to work out what the default payment method should be on the server and send this to the client. 

Please also consider in your review how this work might affect integration with other payment methods or checkout integrations. I've tested with a few payment methods and found no issues so far, but you may know more about any potential hacky workarounds people have done (like in the recent SIRT) or otherwise.

alexflorisca commented on plugins/woocommerce-blocks/assets/js/data/payment/default-state.ts:
```diff
@@ -14,6 +14,79 @@ import {
 import { SavedPaymentMethod } from './types';
 import { STATUS as PAYMENT_STATUS } from './constants';
 
+function getDefaultPaymentMethod() {
```
This now comes from the server. It can either be a string in the case of a "normal" payment method, or an object when it's a saved payment method

alexflorisca commented on plugins/woocommerce-blocks/assets/js/data/payment/default-state.ts:
```diff
@@ -14,6 +14,79 @@ import {
+ * Either as part of the `defaultPaymentMethod` object or
+ * as a token stored in `wcSettings`.
+ */
+function getDefaultPaymentMethodData() {
```
We need this when the default payment method is a saved payment method. It's either extracted from `defaultPaymentMethod` when it's an object, or from the `defaultPaymentMethodToken` setting

mikejolley replied:
Is this so things are available early before rendering? 

Potentially if its east to review this change could be extracted into a separate PR if it improves the initial state.

alexflorisca replied:
This is so that the persisted payment method is loaded as the initial state, so we don't need to fetch it again from client. If it's a saved payment method it will also need the `paymentMethodData` object to be set.

alexflorisca replied:
Doesn't _have_ to be before render, could be hydrated to the checkout store - but that maps to the checkout storeApi endpoint, so I thought this was a better approach

alexflorisca commented on plugins/woocommerce-blocks/assets/js/data/shared-controls.ts:
```diff
@@ -126,7 +126,10 @@ export const apiFetchWithHeadersControl = ( options: APIFetchOptions ) =>
-const preventBatching = [ '/wc/store/v1/cart/select-shipping-rate' ];
+const preventBatching = [
+	'/wc/store/v1/cart/select-shipping-rate',
+	'/wc/store/v1/checkout?calc_totals=true',
```
Prevent batching on the PUT checkout endpoint so that we can cancel it. Batch uses an external library which from what I can tell doesn't support cancelling requests.

mikejolley replied:
Should we just avoid batching on all `/wc/store/v1/checkout` endpoints?

alexflorisca replied:
Yep good idea, done in https://github.com/woocommerce/woocommerce/pull/53553/commits/c5da2ce2286155a5714284ef25b99aa6f12ed79d

alexflorisca replied:
I tried to add a `*` on the end but it doesn't look like it accepts wildcards. The `calc_totals=true` is to avoid expensive calculations by default as per Nadir's comment https://github.com/woocommerce/woocommerce/pull/53553#pullrequestreview-2491763308

alexflorisca commented on plugins/woocommerce/src/StoreApi/Schemas/V1/CheckoutSchema.php:
```diff
@@ -392,7 +407,12 @@ public function validate_additional_fields( $fields, $request ) {
 
-		// Loop over the schema instead of the fields. This is to ensure missing fields are validated.
+		// for PUT requests, we only want to validate the fields that are being updated.
+		if ( $request->get_method() === 'PUT' ) {
```
We only want to validate the supplied fields for PUT requests

Commit:
Fix test error

Commit:
Merge branch 'trunk' into add/42079-additional-fields=persistent-updates

mikejolley commented on plugins/woocommerce-blocks/assets/js/base/context/providers/cart-checkout/checkout-processor.ts:
```diff
@@ -267,6 +268,8 @@ const CheckoutProcessor = () => {
 			...paymentData,
 		};
 
+		clearCheckoutPutRequests();
```
I think for future reference I'd appreciate a comment here as to why its done before fetch.

alexflorisca replied:
Done in https://github.com/woocommerce/woocommerce/pull/53553/commits/ac7cee0845d82dbeefd3cc5858d49572a5c107ae

mikejolley commented on plugins/woocommerce-blocks/assets/js/data/checkout/push-changes.ts:
```diff
@@ -0,0 +1,138 @@
+	};
+
+	// Do we need to push anything?
+	if ( isShallowEqual( localState.checkoutData, newCheckoutData ) ) {
```
This needs to be a fast deep equal because `checkoutData` contains an object (`additionalFields`). Shallow comparison won't work here. _This is a bug._

alexflorisca replied:
I've removed this completely because just below we check if any additional fields or order notes have changed, and return if not

mikejolley commented on plugins/woocommerce/src/StoreApi/Routes/V1/Checkout.php:
```diff
@@ -224,6 +263,58 @@ public function validate_required_additional_fields( \WP_REST_Request $request )
+		}
+
+		/**
+		 * Validate additional fields on request.
```
Is this comment supposed to be here?

alexflorisca replied:
No, good spot, deleted

mikejolley reviewed (changes requested):
Left some comments and questions. There is a bug with shallow compare. I've done code review but no thorough testing. Thanks.

Commit:
prevent batching on all checkout routes

Commit:
Add comment to explain clearCheckoutPutRequests()

Commit:
remove unecessary shallow render check of checkout data

Commit:
Fix don't batch checkout endpoints

Commit:
Remove extra comment

Commit:
Merge branch 'trunk' into add/42079-additional-fields=persistent-updates

Commit:
Sync customer additional fields with customer when persisting so that they don't get overriden on next page load

alexflorisca commented:
I think I've addressed all your feedback @mikejolley, take another look when you can. I'm aware I haven't updated the docs in this PR but it's getting chunky as is so I will do that in a follow up PR as there's quite a few places to change I think.

mikejolley commented:
@alexflorisca there is failing CI. Not sure if flakey or valid.,

Commit:
Check for empty fields before persisting

Commit:
Merge branch 'trunk' into add/42079-additional-fields=persistent-updates

Commit:
fix linting issue

Commit:
Add isCheckoutBlock setting to remove the extra put request on cart block

Commit:
Fix lint issue

alexflorisca commented:
Ok this is ready for another round of review. Fixed the linting in https://github.com/woocommerce/woocommerce/pull/53553/commits/9e1f64954b9b7f1dd4b364eaef7d9327483d6a25. Fixed the extra request on cart by adding `isCheckoutBlock` setting and only fetching when there is a checkout block on the page.

Commit:
Fix totals not updating

Commit:
Merge branch 'trunk' into add/42079-additional-fields=persistent-updates

senadir commented on plugins/woocommerce-blocks/assets/js/base/utils/debounce.ts:
```diff
@@ -30,5 +30,12 @@ export const debounce = < T extends ( ...args: any[] ) => any >(
 		}
 	};
 
+	debounced.clear = () => {
+		if ( timeout ) {
+			clearTimeout( timeout );
+		}
+		timeout = null;
+	};
+
```
Probably worth a comment on why `flush` isn't enough here.

alexflorisca replied:
Done in https://github.com/woocommerce/woocommerce/pull/53553/commits/bc2961dfe4ae8f5b10428901171435432607c37d

senadir commented on plugins/woocommerce-blocks/assets/js/base/utils/debounce.ts:
```diff
@@ -30,5 +30,12 @@ export const debounce = < T extends ( ...args: any[] ) => any >(
 		}
 	};
 
```
```suggestion
	// Flush will always excute pending callback, clear will remove it.
```

alexflorisca replied:
Ah i missed this but I added a similar comment

senadir commented on plugins/woocommerce-blocks/assets/js/data/checkout/push-changes.ts:
```diff
@@ -0,0 +1,132 @@
+ * Initializes the checkout data cache on the first run.
+ */
+const initialize = () => {
+	const store = select( STORE_KEY );
```
We can probably just eliminate the payment `push-changes` and access all data here.

alexflorisca replied:
Done in https://github.com/woocommerce/woocommerce/pull/53553/commits/8a987928e92f9b6b840d21891e7b7d69ad31def4

mikejolley commented on plugins/woocommerce-blocks/assets/js/data/checkout/push-changes.ts:
```diff
@@ -0,0 +1,132 @@
+	localState.checkoutData = {
+		orderNotes: store.getOrderNotes(),
+		additionalFields: store.getAdditionalFields(),
+	};
```
I see you've done a push per data store but I wanted to ask if you think it would be better to have a single checkout push, because right now you'll get two PUT requests if you update notes/fields/payment methods, since payment methods push separately. This might not be necessary since they all PUT to the same endpoint?

alexflorisca replied:
Done in  https://github.com/woocommerce/woocommerce/pull/53553/commits/732c1b846ac9bb5124b40f9281721e22d17e4fd7

senadir commented on plugins/woocommerce-blocks/assets/js/data/utils/clear-put-requests.ts:
```diff
@@ -0,0 +1,17 @@
+import { clearChanges as clearPaymentChanges } from '../payment/push-changes';
+
+// This is used to abort the checkout PUT requests
+export const CheckoutPutAbortController = new AbortController();
```
This needs to be recreated again once aborted.

senadir replied:
This is not a blocker BTW

alexflorisca replied:
Done in https://github.com/woocommerce/woocommerce/pull/53553/commits/73118e652dff1c936f54f4e4b41438022165317a

mikejolley commented on plugins/woocommerce-blocks/assets/js/data/payment/default-state.ts:
```diff
@@ -14,6 +14,79 @@ import {
 import { SavedPaymentMethod } from './types';
 import { STATUS as PAYMENT_STATUS } from './constants';
 
+function getDefaultPaymentMethod() {
```
This should be typed so we can see whats returned (comment says object or string?)

alexflorisca replied:
This has been removed in https://github.com/woocommerce/woocommerce/pull/53553/commits/97978e387cac644a46fd80ea1cc8ee685a40b188

mikejolley commented on plugins/woocommerce-blocks/assets/js/data/payment/index.ts:
```diff
@@ -25,4 +26,6 @@ export const config = {
 export const store = createReduxStore( STORE_KEY, config );
 register( store );
 
+subscribe( pushChanges, store );
```
If we combine pushChanges, we can remove this subscribe.

alexflorisca replied:
We still need to subscribe, but it's now using the push-changes from the checkout store instead

senadir commented on plugins/woocommerce-blocks/assets/js/data/payment/default-state.ts:
```diff
@@ -14,6 +14,79 @@ import {
 
+function getDefaultPaymentMethod() {
+	const defaultPaymentMethod = getSetting< SavedPaymentMethod | string >(
+		'defaultPaymentMethod',
```
```suggestion
		'_defaultPaymentMethod',
```

alexflorisca replied:
This has been removed in https://github.com/woocommerce/woocommerce/pull/53553/commits/97978e387cac644a46fd80ea1cc8ee685a40b188

senadir commented on plugins/woocommerce-blocks/assets/js/data/payment/default-state.ts:
```diff
@@ -14,6 +14,79 @@ import {
 import { STATUS as PAYMENT_STATUS } from './constants';
 
+function getDefaultPaymentMethod() {
+	const defaultPaymentMethod = getSetting< SavedPaymentMethod | string >(
```
As we talked in our sync, ideally in a **followup**, we should move to only read this from the GET request.

mikejolley commented on plugins/woocommerce/src/StoreApi/Routes/V1/Checkout.php:
```diff
@@ -224,6 +263,54 @@ public function validate_required_additional_fields( \WP_REST_Request $request )
+		$this->persist_order_notes_for_order( $request );
+		$this->persist_payment_method_for_order( $request );
+
+		if ( $request->get_param( 'calc_totals' ) ) {
```
Possibly needs experimental prefix while in flux.

alexflorisca replied:
Done in [Add experimental prefix to calc_totals](https://github.com/woocommerce/woocommerce/pull/53553/commits/a1a581ac26e2067acb1315eadcf85afea5ed72f9)

senadir commented on plugins/woocommerce/src/Blocks/BlockTypes/Checkout.php:
```diff
@@ -429,6 +429,7 @@ protected function enqueue_data( array $attributes = [] ) {
 		$this->asset_data_registry->add( 'hasDarkEditorStyleSupport', current_theme_supports( 'dark-editor-style' ) );
 		$this->asset_data_registry->register_page_id( isset( $attributes['cartPageId'] ) ? $attributes['cartPageId'] : 0 );
 		$this->asset_data_registry->add( 'isBlockTheme', wc_current_theme_is_fse_theme() );
+		$this->asset_data_registry->add( 'isCheckoutBlock', true );
```
Once we move all payment logic to PHP let's remove this.

alexflorisca replied:
We can't move this as the checkout store is still loaded on the cart page (should it?) so we need a way to distinguish if we're on the cart or checkout block

mikejolley commented on plugins/woocommerce/src/StoreApi/Schemas/V1/CheckoutSchema.php:
```diff
@@ -201,34 +210,40 @@ public function get_properties() {
+			'billing_address'    => (object) $this->billing_address_schema->get_item_response( $order ),
+			'shipping_address'   => (object) $this->shipping_address_schema->get_item_response( $order ),
+			'payment_method'     => $order->get_payment_method(),
+			'payment_result'     => $payment_result,
```
`null` handling is needed—update schema to show it can be null, and correct the description which says it can be `false`.

alexflorisca replied:
Done in https://github.com/woocommerce/woocommerce/pull/53553/commits/afbdcacf7ff4951291b9a0726358f6da7ac88c4e

senadir commented on plugins/woocommerce/src/StoreApi/Schemas/V1/CheckoutSchema.php:
```diff
@@ -201,34 +210,40 @@ public function get_properties() {
+			'payment_status'  => $payment_result->status,
+			'payment_details' => $this->prepare_payment_details_for_response( $payment_result->payment_details ),
+			'redirect_url'    => $payment_result->redirect_url,
+		] : null;
```
Also add this to schema since payment_result in schema is always an object.

alexflorisca replied:
Done in https://github.com/woocommerce/woocommerce/pull/53553/commits/afbdcacf7ff4951291b9a0726358f6da7ac88c4e

mikejolley commented on plugins/woocommerce/src/StoreApi/Utilities/CheckoutTrait.php:
```diff
@@ -220,5 +225,16 @@ private function persist_additional_fields_for_order( \WP_REST_Request $request
+
+		// If the user is logged in, we need to sync the customer additional fields with the order
+		// otherwise they will be overwritten on next page load.
+		if ( is_user_logged_in() && 0 !== $this->order->get_customer_id() && get_current_user_id() === $this->order->get_customer_id() ) {
```
I don't think you need the first `is_user_logged_in() && ` because the other statements take care of this.

alexflorisca replied:
Done in https://github.com/woocommerce/woocommerce/pull/53553/commits/405c6813fd1607c2cacff4d08842299abf87c112

mikejolley reviewed (commented):
Some feedback after our huddle—let's sync again later this week.

senadir commented on plugins/woocommerce/src/StoreApi/Utilities/CheckoutTrait.php:
```diff
@@ -220,5 +225,16 @@ private function persist_additional_fields_for_order( \WP_REST_Request $request
+
+		// If the user is logged in, we need to sync the customer additional fields with the order
+		// otherwise they will be overwritten on next page load.
+		if ( is_user_logged_in() && 0 !== $this->order->get_customer_id() && get_current_user_id() === $this->order->get_customer_id() ) {
```
This should:

- Be handled on additional fields class level.
- Should not care about if you're logged in or not. Both should be treated the same.

alexflorisca replied:
Done in https://github.com/woocommerce/woocommerce/pull/53553/commits/405c6813fd1607c2cacff4d08842299abf87c112

senadir commented on plugins/woocommerce/src/StoreApi/Routes/V1/Checkout.php:
```diff
@@ -224,6 +263,54 @@ public function validate_required_additional_fields( \WP_REST_Request $request )
+		$this->persist_order_notes_for_order( $request );
+		$this->persist_payment_method_for_order( $request );
+
+		if ( $request->get_param( 'calc_totals' ) ) {
```
Make this variable prefixed so we're not committed to it.

alexflorisca replied:
Done in https://github.com/woocommerce/woocommerce/pull/53553/commits/a1a581ac26e2067acb1315eadcf85afea5ed72f9

senadir commented on plugins/woocommerce/src/Blocks/BlockTypes/Checkout.php:
```diff
@@ -565,34 +567,90 @@ function ( $payment_gateway ) {
+	/**
+	 * Hydrate the default payment method.
+	 */
+	protected function hydrate_default_payment_method() {
```
All of this logic should be moved to Store API responses, on `create_order_from_cart`, you should be setting `$order->set_payment_method( $this->get_default_payment_method() )`

alexflorisca replied:
Done in https://github.com/woocommerce/woocommerce/pull/53553/commits/97978e387cac644a46fd80ea1cc8ee685a40b188

senadir commented on plugins/woocommerce/src/Blocks/BlockTypes/Checkout.php:
```diff
@@ -565,34 +567,90 @@ function ( $payment_gateway ) {
+	 */
+	protected function hydrate_default_payment_method() {
+		$chosen_payment_method       = WC()->session->get( 'chosen_payment_method' );
+		$chosen_payment_method_token = WC()->session->get( 'chosen_payment_method_token' );
```
For tokens let's remove this new session key and just lean in on what's in core

https://github.com/woocommerce/woocommerce//blob/d05b77e66bd35231cd2c82fd87c96ce4f1dba970/docs/payments/payment-token-api.md#L521-L530

alexflorisca replied:
Done in https://github.com/woocommerce/woocommerce/pull/53553/commits/97978e387cac644a46fd80ea1cc8ee685a40b188

senadir reviewed (changes requested):
Thank you @alexflorisca and @mikejolley we reviewed this over huddle and those are the notes.

Commit:
Combine payment and checkout PUT

Commit:
remove payment push-changes

Commit:
Delete updatePaymentMethodData

Commit:
Add experimental prefix to calc_totals

Commit:
Remove push-changes from clear-put-requests

Commit:
Update schema to accept null

Commit:
Remove extra loggeed in check from persist_additional_fields_for_rder

Commit:
Comment debounce

Commit:
Recreate The Abort Controller after it has been used

Commit:
reset checkout.php

Commit:
Move the defaultPaymentMethod to the storeApi

Commit:
Remove payment_method_token and always return the default saved payment method

Commit:
Merge branch 'trunk' into add/42079-additional-fields=persistent-updates

Commit:
Fix enabled_payment_gateways reference

opr reviewed (commented):
Starting to update some docs that will need to change following this, and wondering how the following works:

If a developer wants to do something based on the value of an additional field, how can they react to changes during a PUT? In POST we can use the`woocommerce_store_api_checkout_update_order_from_request` action from `update_order_from_request` but it seems this isn't called on PUT? Looking at the method, it seems there's some shared logic in `get_route_update_response` and `update_order_from_request` - could we just call `update_order_from_request` instead of doing:

```php
$this->persist_additional_fields_for_order( $request );
$this->persist_order_notes_for_order( $request );
$this->persist_payment_method_for_order( $request );
```

in `get_route_update_response`?

mikejolley commented on plugins/woocommerce-blocks/assets/js/data/checkout/push-changes.ts:
```diff
@@ -0,0 +1,173 @@
+	},
+};
+
+const isCheckoutBlock = getSetting< boolean >( 'isCheckoutBlock', false );
```
What property triggers push when we're not on a checkout page? Just trying to understand this part; it looks like a strange workaround and I'd like to understand why its in place.

alexflorisca replied:
The defaultActivePaymentMethod is set only on the Checkout route, not Cart. So on the Cart block, initially the activePaymentMethod is empty, which triggers an update from https://github.com/woocommerce/woocommerce/blob/add/42079-additional-fields%3Dpersistent-updates/plugins/woocommerce-blocks/assets/js/data/payment/utils/set-default-payment-method.ts#L24

alexflorisca replied:
I think there's been some discussion around having some way to tell if we're on the cart or checkout page client side as well, so I think this will be useful

mikejolley commented on plugins/woocommerce-blocks/assets/js/data/checkout/push-changes.ts:
```diff
@@ -0,0 +1,173 @@
+		orderNotes: store.getOrderNotes(),
+		additionalFields: store.getAdditionalFields(),
+		activePaymentMethod: paymentStore.getActivePaymentMethod(),
+	};
```
👍🏻

mikejolley commented on plugins/woocommerce/src/StoreApi/Utilities/PaymentUtils.php:
```diff
@@ -0,0 +1,113 @@
+		 * @param \WC_Token $token     The token for the current list item.
+		 *
+		 * @return array The list item with the token id added.
+		 */
```
Spacing is off here

mikejolley commented on plugins/woocommerce/src/StoreApi/Utilities/PaymentUtils.php:
```diff
@@ -0,0 +1,113 @@
+		$brand                = ! empty( $list_item['method']['brand'] ) ?
+			strtolower( $list_item['method']['brand'] ) :
+			'';
+		// phpcs:ignore WordPress.WP.I18n.TextDomainMismatch -- need to match on translated value from core.
```
```suggestion
```

Not needed. We use this textdomain now.

mikejolley reviewed (commented):
Thanks for working on those changes. Looking easier to understand now. Added a few questions.

Commit:
Update plugins/woocommerce/src/StoreApi/Utilities/PaymentUtils.php

Co-authored-by: Mike Jolley <mike.jolley@me.com>

Commit:
Fix spacing on paymentutils

Commit:
Merge branch 'add/42079-additional-fields=persistent-updates' of github.com:woocommerce/woocommerce into add/42079-additional-fields=persistent-updates

Commit:
Merge branch 'trunk' into add/42079-additional-fields=persistent-updates

Commit:
Fix linting

Commit:
Fix tests

Commit:
Merge branch 'trunk' into add/42079-additional-fields=persistent-updates

Commit:
Fix actions

Commit:
Merge branch 'trunk' into add/42079-additional-fields=persistent-updates

Commit:
Fix bug where payment methods are null

Commit:
Change gov-id sto student-id in test to avoid possible conflict with other tests

Commit:
Merge branch 'trunk' into add/42079-additional-fields=persistent-updates

wavvves committed:
Changed assert type for PHP compat

wavvves committed:
Fix bad trunk conflict merge.

wavvves committed:
Unnecessary var removal.

wavvves committed:
Lint 😖

mikejolley commented on plugins/woocommerce/src/StoreApi/Routes/V1/Checkout.php:
```diff
@@ -112,6 +112,28 @@ public function get_args() {
+							'type'        => 'object',
+						],
+						'payment_method'    => [
+							'description' => __( 'Seleted payment method for the order.', 'woocommerce' ),
```
```suggestion
							'description' => __( 'Selected payment method for the order.', 'woocommerce' ),
```

mikejolley reviewed (commented):
I've done testing with Stripe and Fields plugins, aside from express methods. Found one issue as discussed with the extra request on page load. Will check again once patched.

There is also one typo. Ill approve once thats resolved. Thanks.

Commit:
Fix extra request on load for checkout

Commit:
Update plugins/woocommerce/src/StoreApi/Routes/V1/Checkout.php

Co-authored-by: Mike Jolley <mike.jolley@me.com>

Commit:
Merge branch 'trunk' into add/42079-additional-fields=persistent-updates

mikejolley reviewed (approved):
Thanks its working now. If tests continue to pass I think we're good 👍🏻

senadir reviewed (approved):
Thank you Alex, I reviewed your latest changes and they made sense to me. We just need to be more vigilant when merging additional fields work on other branches.
